src/client_impl.cc

```diff
@@ -98,21 +98,26 @@ void MixerClientImpl::Check(const Attributes &attributes,
   if (!transport) {
     transport = options_.check_transport;
   }
-  transport(request, response, [response, raw_check_result, raw_quota_result,
-                                on_done](const Status &status) {
-    raw_check_result->SetResponse(status, *response);
-    raw_quota_result->SetResponse(status, *response);
-    if (on_done) {
-      if (!raw_check_result->status().ok()) {
-        on_done(raw_check_result->status());
-      } else {
-        on_done(raw_quota_result->status());
-      }
-    }
-    delete raw_check_result;
-    delete raw_quota_result;
-    delete response;
-  });
+  transport(request, response,
+            [this, response, raw_check_result, raw_quota_result,
+             on_done](const Status &status) {
+              raw_check_result->SetResponse(status, *response);
+              raw_quota_result->SetResponse(status, *response);
+              if (on_done) {
+                if (!raw_check_result->status().ok()) {
+                  on_done(raw_check_result->status());
+                } else {
+                  on_done(raw_quota_result->status());
+                }
+              }
+              delete raw_check_result;
+              delete raw_quota_result;
+              delete response;
+
+              if (InvalidDictionaryStatus(status)) {
```

Comment:
So if dictionary mismatch is treated as transport failure, the place to check this should be before call `on_done`, since in that case, your `response` will be a default protobuf instance, (i.e. status is OK), so the `on_done` is called with success status while it is not.